### PR TITLE
Move synthesis code into repo

### PR DESCRIPTION
**Pull request by D0048**
A primitive PoC pipeline training a segmentation network with synthetic Infantry placed on Imagenet background.

### CODE REVIEW

File: dnn/POC_synthetic_segmentation/generation/scene_setup_transparent.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+            loc[(loc>=0)*(loc<=2)]=-2
+            obj.location.x,obj.location.y,obj.location.z=tuple(loc)
+            obj.keyframe_insert('location', frame=i)
+        
+        for l in light_objects:
+            l.data.node_tree.nodes['Emission'].inputs['Strength'].default_value=np.random.rand()*20
+            l.data.node_tree.nodes['Emission'].inputs['Strength'].keyframe_insert(
+                data_path='default_value', frame=i)
+                
+    #bpy.app.handlers.frame_change_post.remove(update)
```
Comment (alvinsunyixiao) [inline dnn/POC_synthetic_segmentation/generation/scene_setup_transparent.py:92]: dead code?

File: dnn/POC_synthetic_segmentation/generation/scene_setup_transparent.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+if __name__ == '__main__':
+    scene = bpy.context.scene
+    obj_camera = bpy.data.objects["Camera"]
+    robot_objects = [obj for obj in bpy.context.scene.collection.children.get('Robot').objects]
+    label_objects=[obj for obj in bpy.context.scene.collection.children.get('Label').objects]
+    stage_objects=[obj for obj in bpy.context.scene.collection.children.get('Environment').objects]
+    light_objects=[obj for obj in bpy.context.scene.collection.children.get('Light').objects]
+    number_objects=[obj for obj in bpy.context.scene.collection.children.get('Number').objects]
+    #set_visible_all(robot_objects,visibility=False)
```
Comment (alvinsunyixiao) [inline dnn/POC_synthetic_segmentation/generation/scene_setup_transparent.py:105]: Why are these commented?

File: dnn/POC_synthetic_segmentation/training/fcn.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+
+            nn.Upsample([res[1],res[0]]),
+            nn.Conv2d(in_channels=channels, out_channels=out_channels,
+                      kernel_size=kernel_size, stride=1, padding=padding),
+            nn.Sigmoid(), ]
+        self.ops = nn.ModuleList(ops)
+
+    def forward(self, x):
+        # print(x.shape)
```
Comment (alvinsunyixiao) [inline dnn/POC_synthetic_segmentation/training/fcn.py:61]: remove debug prints

File: dnn/POC_synthetic_segmentation/training/fcn.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        for op in self.ops:
+            x = op(x)
+            # print(x.shape)
```
Comment (alvinsunyixiao) [inline dnn/POC_synthetic_segmentation/training/fcn.py:64]: ditto

File: dnn/POC_synthetic_segmentation/training/train.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
+from img_utils import *
+#from unet import *
```
Comment (alvinsunyixiao) [inline dnn/POC_synthetic_segmentation/training/train.py:2]: kill unused imports

File: dnn/POC_synthetic_segmentation/training/train.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+for i in range(16):
+    p1 = torch.multiprocessing.Process(target=load_train, args=(q_train,))
+    p1.start()
+
+save_directory_name = './checkpoint'
+sample_directory_name = './samples'
+
+optimizer = torch.optim.Adam(model.parameters(), lr=1.0e-4)
+# optimizer = torch.optim.RMSprop(model.parameters(), lr=1.0e-3)
+# criterion = nn.BCELoss().to(device)
```
Comment (alvinsunyixiao) [inline dnn/POC_synthetic_segmentation/training/train.py:61]: If experiment says Adam with MSE loss is better, then maybe remove those commented settings?

File: dnn/POC_synthetic_segmentation/training/unet.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from img_utils import *
+
+
+def downconv(in_channels, out_channels, kernel_size):
+    padding = int(kernel_size / 2)
```
Comment (alvinsunyixiao) [inline dnn/POC_synthetic_segmentation/training/unet.py:8]: nit
```suggestion
    padding = kernel_size // 2
```

File: dnn/POC_synthetic_segmentation/generation/scene_setup_transparent.py [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+        obj_camera.data.show_background_images=False
+    else:
+        scene.render.filepath = bpy.data.filepath.rpartition('.')[0]+'/image_out/out_seed{}_'.format(seed)
+        obj_camera.data.show_background_images=True
+    set_visible_all(get_all_objs(),not label)
+    set_visible_all(label_objects,label)
+    
+    try: 
+        bpy.app.handlers.frame_change_post.remove(update)
+    except: pass
```
Comment (alvinsunyixiao) [inline dnn/POC_synthetic_segmentation/generation/scene_setup_transparent.py:61]: What is this doing?

**Review by alvinsunyixiao — COMMENTED** [review]
LGTM. Just some scattered dead code that is probably better removed.

File: dnn/POC_synthetic_segmentation/generation/scene_setup_transparent.py (hunk 8 above)
Comment (D0048) [reply dnn/POC_synthetic_segmentation/generation/scene_setup_transparent.py:62]: Re-registering frame update callback with the newer update function in code. Useful when you modify the update function and rerun the script without restarting blender.

**Comment by D0048** [comment]
Dead-code are now removed.

File: dnn/POC_synthetic_segmentation/generation/scene_setup_transparent.py (hunk 8 above)
Comment (alvinsunyixiao) [reply dnn/POC_synthetic_segmentation/generation/scene_setup_transparent.py:62]: Why do we need this `try` `except: pass` thing? The error handling looks a bit sketchy

Comment (D0048) [reply dnn/POC_synthetic_segmentation/generation/scene_setup_transparent.py]: I agree. Changed logic to just clear all old callbacks instead.

**Comment by D0048** [comment]
`LCMUtilsTest.LCMDecodeTest` failed although no changes to related code. Might worth a separate issue if it repeats?

**Review by alvinsunyixiao — APPROVED** [review]
Cool stuff. LGTM. 

Yeah about LCM Decoder test, it is already an open issue. Let's fix that in a separate review
